File: cpp/1105.cpp
#include<iostream>
#include<vector>
#include<climits>

using namespace std;

class Solution{
public:
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth){
        int n = books.size();
        vector<int> dp(n+1, INT_MAX);
        dp[0] = 0;
        for(int i=1; i<= n; i++){
            int width = 0, height = 0;
            for(int j=i; j>0; j--){
                width += books[j-1][0];
                if(width > shelfWidth)
                    break;
                height = max(height, books[j-1][1]);
                dp[i] = min(dp[i], dp[j-1] + height);
            }
        }
        return dp[n];
    }
};

int main(){
    int n;
    cin >> n;
    vector<vector<int>> books(n, vector<int>(2));
    for(int i = 0; i < n; i++)
        cin >> books[i][0] >> books[i][1];
    int shelfWidth;
    cin >> shelfWidth;
    Solution s;
    cout << s.minHeightShelves(books, shelfWidth) << endl;
    return 0;
}

File: python/1105.py
class Solution:
    def minHeightShelves(self, books: list[list[int]], shelfWidth: int) -> int:
        n = len(books)
        dp = [float('inf')] * (n+1)
        dp[0] = 0
        for i in range(1, n+1):
            width, height = 0, 0
            j = i
            while j > 0:
                width += books[j-1][0]
                if width > shelfWidth:
                    break
                height = max(height, books[j-1][1])
                dp[i] = min(dp[i], dp[j-1] + height)
                j -= 1
        return dp[n]
    
if __name__ == '__main__':
    n = int(input())
    books = [list(map(int, input().split())) for _ in range(n)]
    shelfWidth = int(input())
    s = Solution()
    print(s.minHeightShelves(books, shelfWidth))
            